### ISSUE TITLE
`OSSL_SIGNATURE_PARAM_ALGORITHM_ID` insufficiently documented and hard to use by provider authors

### ISSUE THREAD

**Issue by DDvO** [openssl/openssl#22932]
When applications, including the OpenSSL apps, use the convenient function `ASN1_item_sign{,_ex()}` and friends for signing ASN.1 data, `ASN1_item_sign_ctx()` defined in `crypto/asn1/a_sign.c` is called, which may also be called directly.

These three functions have _two_ parameters `X509_ALGOR *algor1` and `X509_ALGOR *algor2`, which are just very vaguely documented as:

> The ASN1_item_sign_ex() function will write the correct X509_ALGOR
> to the structs based on the algorithms and parameters that have been set up. If one of
> algor1 or algor2 points to the X509_ALGOR of the data to be signed, then that X509_ALGOR
> will first be written before the signature is generated. 

It remains unclear why two such arguments can be given. At first I thought this is because signing usually involves an asymmetric signature alg and a digest (hash) alg, but if both are present, they get combined in a single `X509_ALGOR`.
So I guess the reason is to cater for potential future hybrid signatures for post-quantum transition?

When a crypto provider with signature capabilities is involved, `ASN1_item_sign_ctx()` queries it and processes the result like this:
```
        params[0] =
            OSSL_PARAM_construct_octet_string(OSSL_SIGNATURE_PARAM_ALGORITHM_ID,
                                              aid, sizeof(aid));
        params[1] = OSSL_PARAM_construct_end();

        if (EVP_PKEY_CTX_get_params(pctx, params) <= 0)
            goto err;

        if ((aid_len = params[0].return_size) == 0) {
            ERR_raise(ERR_LIB_ASN1, ASN1_R_DIGEST_AND_KEY_TYPE_NOT_SUPPORTED);
            goto err;
        }

        if (algor1 != NULL) {
            const unsigned char *pp = aid;

            if (d2i_X509_ALGOR(&algor1, &pp, aid_len) == NULL) {
                ERR_raise(ERR_LIB_ASN1, ERR_R_INTERNAL_ERROR);
                goto err;
            }
        }

        if (algor2 != NULL) {
            const unsigned char *pp = aid;

            if (d2i_X509_ALGOR(&algor2, &pp, aid_len) == NULL) {
                ERR_raise(ERR_LIB_ASN1, ERR_R_INTERNAL_ERROR);
                goto err;
            }
        }
```

There are multiple issues with this.

1. Provider authors need to be aware of the requirement to provide a response to the `OSSL_SIGNATURE_PARAM_ALGORITHM_ID` query when `ASN1_item_sign_ex()` is involved - this is not documented. `provider-signatures.pod` just says:
> Gets the DER encoded AlgorithmIdentifier that corresponds to the combination of
> signature algorithm and digest algorithm for the signature operation

~This does not mention that potentially two such encodings can be given, which then need to be concatenated.~

2. It is misleading that `provider-signature.pod` says:
> OSSL_FUNC_signature_set_ctx_params and OSSL_FUNC_signature_settable_ctx_params are optional,
> but if one of them is present then the other one must also be present. The same applies to
> OSSL_FUNC_signature_get_ctx_params and OSSL_FUNC_signature_gettable_ctx_params, as well as
> the "md_params" functions. The OSSL_FUNC_signature_dupctx function is optional.

because at least when `ASN1_item_sign_ex()` is involved, `OSSL_FUNC_signature_get_ctx_params` and `OSSL_FUNC_signature_gettable_ctx_params` must be present - otherwise, more or less obscure/misleading errors are thrown (if at all) about algorithms not being supported.

3. Once a provider author has found out that (s)he has to implement those functions and the `OSSL_SIGNATURE_PARAM_ALGORITHM_ID` query, (s)he faces the challenge to produce that DER-encoded ASN.1 AlgorithmIdentifier, which can be painful. Some even hand-craft ASN.1 encodings (rather than using `i2d_X509_ALGOR()`) - here is a deterring example: the stretch https://github.com/latchset/pkcs11-provider/blob/main/src/signature.c#L204 to https://github.com/latchset/pkcs11-provider/blob/main/src/signature.c#L363 and some further code below there.
This ASN.1 encoding appears pretty needless because `ASN1_item_sign_ctx()` and the two other consumers within OpenSSL (all in `crypto/`) immediately parse and discard it. Yet it looks like there is a fundamental design decision/requirement that parameters need to be serialized in a universally parseable form - quoting from `OSSL_PARAM.pod`:
> B<OSSL_PARAM> is a type that allows passing arbitrary data for some
> object between two parties that have no or very little shared
> knowledge about their respective internal structures for that object.

Solution options that I see:

1. Provide the providers of providers ;-) with a helper function that produces the wanted DER-encoded AlgorithmIdentifier ASN.1 structure from the signature and optional digest algorithm names. This could be done at least for the vast majority of use cases, possibly not supporting less common options (such as RSA-PKCS1-PSS).
2. Rather than asking the provider via `OSSL_SIGNATURE_PARAM_ALGORITHM_ID` to do so, change the consuming functions `ASN1_item_sign_ctx()` etc. to directly construct the wanted `X509_ALGOR` structure(s) itself from the chosen signature and (optional) digest algorithms, at least for the most common use cases.
`rsa_item_sign()` and `rsa_cms_sign()` already do so for the typical RSA-PKCS1 case.  
`ASN1_item_sign_ctx()` does not, but can do as it already knows about these algorithms (and in case any other consuming functions do not, they could query the provider for the algorithm names, I suppose), at least for the most common RSA and EC cases.
3. Add an alternative parameter interface that does not use ASN.1, which could be queried as a fallback in case the provider does not give an answer to `OSSL_SIGNATURE_PARAM_ALGORITHM_ID`.
For instance, it could directly involve one or two pointers to `X509_ALGOR` structures - yet it looks like the `OSSL_PARAM` design does not allow this straightforward approach.

**Comment by mattcaswell** [comment]
> Some even hand-craft ASN.1 encodings

We do this too...well the code to do it is auto-generated. See the various code files here:

https://github.com/openssl/openssl/tree/master/providers/common/der

The reason being is that the `i2d_X509_ALGOR()` function is deliberately not available in the FIPS provider (none of the ASN.1 code is).

There isn't really a big problem with doing this, since the encodings are basically static for any given algorithm.

**Comment by DDvO** [comment]
> > This does not mention that potentially two such encodings can be given, which then need to be concatenated.
> 
> I don't see any concatenation in the above code? It just looks to me like the aid buffer is processed from the beginning twice. Once for algor1 and once for algor2.

Oh, right - yet then what is the sense of parsing the same piece of ASN.1 encoding twice?
I'm pretty sure the result will not become any better for `algor2` ;-)

**Comment by DDvO** [comment]
This issue is just a specific instance of the pretty fundamental issue #22940.

**Comment by DDvO** [comment]
Yes, `algor1` and `algor2` are output parameters, which is not immediately visible, and this is why I added a comment pointing this out, e.g., at https://github.com/openssl/openssl/blob/master/crypto/cmp/cmp_protect.c#L124

> The best I can figure out is that it's completely up to the backend (engine) how those are actually used, the idea being, however, that one of them represents the pkey algo while the other represents the digets algo.

If so, `ASN1_item_sign_ctx()` does not implement this idea.

**Comment by davidben** [comment]
The reason for there being both `algor1` and `algor2` is just because an X.509 certificate has two copies of the signature algorithm, one inside the TBSCertificate and one outside. And then, even though they're required to match, OpenSSL's in-memory representation just preserves the fact that there are two of them. I guess someone, a long time ago, thought it'd be convenient to have the function output twice and that got carried over.

**Comment by DDvO** [comment]
Meanwhile I've tried out several ways of adding to a given provider implementation some code implementing a response for `OSSL_SIGNATURE_PARAM_ALGORITHM_ID`, without spending many days of developing a solution from scratch, but all of them are complicated, partly ugly, and for the most part not working.
1. The horrible [hand-coded ASN1.structure approach](https://github.com/latchset/pkcs11-provider/blob/main/src/signature.c#L204) was at least not too hard to take over as a PoC, and it works.
2. Then I tried copying over `rsa_generate_signature_aid()` etc. from `providers/implementations/signature/*sig.c` but due to the use of many further internal functions and other definitions I had to give up on this.
3. Then I tried copying over `tpm2_sig_scheme_to_x509_alg()`, which is [part of the tpm2-openssl provider](https://github.com/tpm2-software/tpm2-openssl/blob/1555087948b5a50e358cd737cfa8b671dcc2a1ed/src/tpm2-provider-x509.c#L221). It was not too hard to make it compile, but for some reason it fails/crashes in my context.

So please, do provide a public default/helper implementation for `OSSL_SIGNATURE_PARAM_ALGORITHM_ID`, for instance as a wrapper combining the respective functions in `providers/implementations/signature/*sig.c`.

This issue really is also a feature request.

**Comment by DDvO** [comment]
I meanwhile had a look at this again and provided PR #27372 fixing the documentation part of this issue.
